x/programs/rust/wasmlanche-sdk/src/memory.rs

```diff
@@ -88,22 +84,100 @@ pub extern "C" fn alloc(len: usize) -> HostPtr {
+    ALLOCATIONS
+        .with_borrow_mut(|allocations| allocations.remove(&ptr))
+        .expect("attempted to drop invalid host pointer")
+}
+
 #[cfg(test)]
 mod tests {
+    #![allow(clippy::useless_vec)]
     use super::*;
 
     #[test]
-    fn data_allocation() {
-        let len = 1024;
-        let ptr = alloc(len);
-        let vec: Vec<_> = (u8::MIN..=u8::MAX).cycle().take(len).collect();
-        unsafe { std::ptr::copy(vec.as_ptr(), ptr.0.cast_mut(), vec.len()) }
-        assert_eq!(*ptr, vec);
+    #[should_panic(expected = "attempted to deref invalid host pointer")]
+    fn deref_untracked_pointer() {
+        let ptr = HostPtr(std::ptr::null());
+        let _ = &*ptr;
+    }
 
-        let inner = ptr.0;
+    #[test]
+    #[should_panic(expected = "attempted to drop invalid host pointer")]
+    fn drop_untracked_pointer() {
+        let data = vec![0xff, 0xaa];
+        let ptr = HostPtr(data.as_ptr());
         drop(ptr);
+    }
+
+    #[test]
+    fn deref_tracked_pointer() {
+        let data = vec![0xff];
+        let cloned = data.clone();
+        let data = ManuallyDrop::new(data);
+        let ptr = data.as_ptr();
+
+        ALLOCATIONS.with_borrow_mut(move |allocations| {
+            allocations.insert(ptr, data.len());
+        });
+
+        assert_eq!(&*HostPtr(ptr), &cloned);
+    }
+
+    #[test]
+    fn deref_is_borrow() {
+        let data = vec![0xff];
+        let cloned = data.clone();
+        let data = ManuallyDrop::new(data);
+        let ptr = data.as_ptr();
+
+        ALLOCATIONS.with_borrow_mut(move |allocations| {
+            allocations.insert(ptr, data.len());
+        });
+
+        let host_ptr = ManuallyDrop::new(HostPtr(ptr));
+        assert_eq!(&**host_ptr, &cloned);
+        let host_ptr = HostPtr(ptr);
+        assert_eq!(&*host_ptr, &cloned);
+    }
+
+    #[test]
+    fn host_pointer_to_vec_takes_bytes() {
+        let data = vec![0xff];
+        let cloned = data.clone();
+        let data = ManuallyDrop::new(data);
+        let ptr = data.as_ptr();
+
+        ALLOCATIONS.with_borrow_mut(move |allocations| {
+            allocations.insert(ptr, data.len());
+        });
+
+        assert_eq!(Vec::from(HostPtr(ptr)), cloned);
+
+        ALLOCATIONS.with_borrow(|allocations| {
+            assert!(allocations.get(&ptr).is_none());
+        });
+    }
+
+    #[test]
+    fn dropping_host_pointer_deallocates() {
+        let data = vec![0xff];
+        let data = ManuallyDrop::new(data);
+        let ptr = data.as_ptr();
+
+        ALLOCATIONS.with_borrow_mut(move |allocations| {
+            allocations.insert(ptr, data.len());
+        });
+
+        drop(HostPtr(ptr));
+
+        ALLOCATIONS.with_borrow(|allocations| {
+            assert!(allocations.get(&ptr).is_none());
+        });
 
-        ALLOCATIONS.with_borrow(|map| assert!(!map.contains_key(&inner)));
+        // overwrites old allocation
+        let data = vec![0x00];
+        assert_eq!(data.as_ptr(), ptr);
```

Comment:
why would this always write into the same location?

Reply:
I think there's some level of determinism in the standard allocators. Of course, this definitely isn't guaranteed, but I think it's reasonable to assume that if you're only allocating and deallocating a single byte, an efficient allocator will re-allocate in the same spot.